specs/logging_implementation.adoc

```diff
@@ -0,0 +1,273 @@
+= Generic message mechanism for logs and alerts in Tendrl"
+
+Logs and alerts need to be treated as messages emitted by Tendrl.
+
+== Problem description
+* The logs, alerts and in flight operations updates in Tendrl are handled differently
+  and there is no standard mechanism which makes sure no internal details are
+  exposed as part of alerts/notifications sent to end user.
+* Ideally logs messages, in flight operation updates and alerts are messages which
+  are emitted from an application.
+
+== Use Cases
+
+Logs and alerts need to be treated as messages.
+
+== Proposed change
+* Logs and alerts need to be treated as messages emitted by Tendrl modules.
+* A schema needs to be defined for emitted messages.
+* Message priorities, types etc. need to be defined at the time of emission.
+* The integration/bridge applications must pass the message as per schema.
+* For messages that serve as notifications to the administrator (or alerts),
+  exact error message must be hardcoded by the developers. No backtraces or
+  system generated error messages should be sent as notifications to the
+  administrator.
+* Notifications sent to the administrator should provide sufficient context to
+  enable debugging of the problem via investigation into the log files.
+* Notifications and alerts need to be delivered to the node agent alert sockets.
+* Common module needs to implement a logging library that hides all the implementation
+  details and configuration options from any of the components using the library.
+  Components should be able to blindly emit messages to the Message Handler module
+  in Tendrl and it should be able to deliver them to either syslog, log files,
+  alert sockets etc. depending upon the metadata and configuration.
+* Syslog must be the default logging configuration. Logging to individual
+  configuration files should be thoroughly discouraged.
+
+== Alternatives
+
+* None
+
+== Data model impact
+
+* Store the operation details in etcd:
+```
+1. messages with inorder_id
+    {job_id}/{in_order_id}
+```
+
+== Impacted modules
+
+=== Tendrl/common impact
+* Create a new class called "Message" in common for message handling with attributes
+  handler_name, message_type, message, log_level, timestamp(format: ISO 8601).
+* Create a new class called "Log" in common and inherit it from Message class.
+* Create a new class called "Alert" in common with attributes alert_id, node_id,
+  resource, current_value, tags, type, severity, significance, ackedby, acked.
+  Inherit alert class from Message class.
+* Create a new class called "Operation" in common with attributes
```

Comment:
@brainfunked This job_id and parent_id and flow_id is for currently running job or i have to store all the job in tendrl context